joint_limits/src/simple_joint_limiter.cpp

```diff
@@ -0,0 +1,386 @@
+        auto pos = std::clamp(
+          desired_pos[index], joint_limits_[index].min_position, joint_limits_[index].max_position);
+        if (pos != desired_pos[index])
+        {
+          desired_pos[index] = pos;
+          limited_jnts_pos.emplace_back(joint_names_[index]);
+        }
+        // priority to pos_cmd derivative over cmd_vel because we always have a pos_state so
+        // recomputing vel_cmd is fine compute expected_vel with already clamped pos_cmd and
+        // pos_state
+        // TODO(destogl) handle the case of continuous joints with angle_wraparound to compute vel
+        // correctly
+        desired_vel[index] =
+          (desired_pos[index] - current_joint_states.positions[index]) / dt_seconds;
+      }
+    }
+
+    // limit velocity
+    if (joint_limits_[index].has_velocity_limits)
+    {
+      // clamp input vel_cmd
+      if (std::fabs(desired_vel[index]) > joint_limits_[index].max_velocity)
+      {
+        desired_vel[index] = std::copysign(joint_limits_[index].max_velocity, desired_vel[index]);
+        limited_jnts_vel.emplace_back(joint_names_[index]);
+
+        // recompute pos_cmd if needed
+        if (has_pos_cmd)
+        {
+          desired_pos[index] =
+            current_joint_states.positions[index] + desired_vel[index] * dt_seconds;
+          limited_jnts_pos.emplace_back(joint_names_[index]);
+        }
+
+        // compute desired_acc when velocity is limited
+        desired_acc[index] =
+          (desired_vel[index] - current_joint_states.velocities[index]) / dt_seconds;
+      }
+    }
+
+    // limit acceleration
+    if (
+      joint_limits_[index].has_acceleration_limits || joint_limits_[index].has_deceleration_limits)
+    {
+      // if(has_vel_state)
+      if (1)  // for now use a zero velocity if not provided
+      {
+        // limiting acc or dec function
+        auto apply_acc_or_dec_limit = [&](
+                                        const double max_acc_or_dec, std::vector<double> & acc,
+                                        std::vector<std::string> & limited_jnts) -> bool
+        {
+          if (std::fabs(acc[index]) > max_acc_or_dec)
+          {
+            acc[index] = std::copysign(max_acc_or_dec, acc[index]);
+            limited_jnts.emplace_back(joint_names_[index]);
+            return true;
+          }
+          else
+            return false;
+        };
+
+        // limit acc for pos_cmd and/or vel_cmd
+
+        // compute desired_acc with desired_vel and vel_state
+        desired_acc[index] =
+          (desired_vel[index] - current_joint_states.velocities[index]) / dt_seconds;
+
+        // check if decelerating - if velocity is changing toward 0
+        bool deceleration_limit_applied = false;
+        bool limit_applied = false;
+        if (
+          (desired_acc[index] < 0 && current_joint_states.velocities[index] > 0) ||
+          (desired_acc[index] > 0 && current_joint_states.velocities[index] < 0))
```

Comment:
There is an "outlier" case where it is changing directions i.e. positive to negative velocities or vice-versa, and it will be treated in this logic block as a deceleration when it really is a deceleration and then an acceleration. In fact, if the current velocity is closer to 0 than the desired, it is mostly an acceleration. So I think in this situation, both deceleration and acceleration limits need to be tested. It is a special case that needs to be handled.

Reply:
OK, I understand. When this special case happens, then we have to first consider deceleration limits (later acceleration limits). As we assume that the limiter is running on the control frequency, it should be called multiple times during that period, so the case should be cached. Please confirm if my understanding is correct, maybe I am still missing something.